Feature: Intallation of shopify-cli downloads shopify-extensions binary

dmhenry opened the pull request:
<!--
  ☝️How to write a good PR title:
  - Prefix it with [Feature] (if applicable)
  - Start with a verb, for example: Add, Delete, Improve, Fix…
  - Give as much context as necessary and as little as possible
  - Use a draft PR while it’s a work in progress
-->

### WHY are these changes introduced?

[Enables Development Server](https://github.com/Shopify/shopify-cli-extensions/issues/238) for development of checkout extensions.

### WHAT is this pull request doing?

Installation of the `shopify-cli` gem will trigger download of `shopify-extensions` binary.

https://user-images.githubusercontent.com/4490738/157476748-5a2fadfd-82ed-44f4-8a10-1636c8cb93ac.mp4

### How to test your changes?

1. Navigate to `Shopify/shopify-cli` directory.
2. From the CLI, run `MY_TMP_DIR=$(ruby -e 'puts Gem.dir')` and to save off the target gem installation directory. If using multiple shell sessions you will need to `export MY_TMP_DIR`.
3. Checkout this branch: `git fetch; git checkout origin/checkout-extensions-support`
4. Run `rake install` to build the gem locally to `pkg/`. You can verify this with `ls pkg/`. The next steps assume the output file is `shopify-cli-2.13.0.gem` but adjust your version as needed.
5. Run `gem install --local pkg/shopify-cli-2.13.0.gem` to install the gem built in step 4 to the system directory from step 2 (`MY_TMP_DIR`).
6. Navigate to the local gem installation extension directory: `cd $MY_TMP_DIR/gems/shopify-cli-2.13.0/ext/shopify-extensions`. 
7. Run `./shopify-extensions version` and verify the output is `v0.2.0`
8. Unset the temporary variable: `unset MY_TMP_DIR`

### Post-release steps

<!--
  If changes require post-release steps, for example merging and publishing some documentation changes,
  specify it in this section and add the label "includes-post-release-steps".
  If it doesn't, feel free to remove this section.
-->

### Update checklist

- [ ] I've added a CHANGELOG entry for this PR (if the change is public-facing)
- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows).
- [x] I've left the version number as is (we'll handle incrementing this when releasing).
- [x] I've included any post-release steps in the section above (if needed).

lopert committed:
refactor push package repo wasm module reading

lopert committed:
refactor build errors

lopert committed:
tests

pepicrft committed:
Fix Homebrew template

pepicrft committed:
Update CHANGELOG

Commit:
Intallation of shopify-cli downloads extension binary

dmhenry commented on shopify-cli.gemspec:
```diff
@@ -28,7 +28,7 @@ Gem::Specification.new do |spec|
     %x(git ls-files -z).split("\x0").reject do |f|
       f.match(%r{^(test|spec|features|packaging)/}) ||
-      f.match(%r{^bin/(update-deps|shopify.bat)$})
+        f.match(%r{^bin/(update-deps|shopify.bat)$})
```
Rubocop formatting

t6d commented:
@pepicrft, how do we generally test compatibility with Windows? I tested the binary download a few months ago on Windows, but I think it would be good to repeat that test prior to merging this PR. This is the first time installing the Shopify CLI would cause the download of the binary for every user.

lopert committed:
reviewer feedback

Commit:
Fix `theme serve` when there is a permissions error (with `ThemeAdminAPI` wrapper) (#2114)

pepicrft commented:
> @pepicrft, how do we generally test compatibility with Windows? I tested the binary download a few months ago on Windows, but I think it would be good to repeat that test prior to merging this PR. This is the first time installing the Shopify CLI would cause the download of the binary for every user.

The process I was suggested a while ago was virtualizing locally but it's very cumbersome because you need to set everything up locally. I think the fastest is to set up a GitHub Actions workflow that runs on a Windows server and run the logic there.

dmhenry commented:
@t6d Verified the binary downloads properly in a Linux environment (Spin).

t6d committed:
ShopifyCLI::Result#unwrap! (#2142)

Commit:
Fix ignore file handling in `DevServer::Watcher` (#2133)

t6d committed:
NpmPackage scripts & dependencies (#2134)

NpmPackage now supports parsing a package specification from a stream
and extracts dependencies, devDependencies and scripts in addition to
name and version of a package.

t6d committed:
Make AskTemplate use DevelopmentServerRequirements internally

t6d committed:
Require yaml in DevelopmentServer

t6d committed:
Require pathname in MergeServerConfig

Commit:
Resolve merge conflicts
